### DISCUSSION TITLE
How to Debug Yolov5 model

### DISCUSSION THREAD

**Discussion by Corallo** [AxisCommunications/axis-model-zoo#50] [General]
This informal guide aims to help users debug issues with the YOLOv5 model, obtained following the [yolov5-on-artpec8 guide](https://github.com/AxisCommunications/axis-model-zoo/blob/main/docs/yolov5-on-artpec8.md). The principles can be applied to other networks as well.

Your issue could stem from various sources, such as model architecture patching, training, exporting, output parsing, or Axis OS. To debug effectively, you need to isolate the issue by testing your model step by step.

1. Verify Axis OS Version
Ensure your device runs **Axis OS >= 11.7**, as YOLOv5 is supported starting from this version.

2. Inspect Model Architecture
Use [netron.app](https://netron.app/) to visualize your model (an [offline version](https://github.com/lutzroeder/netron) is available if you prefer not to share your model online).
**Correct architecture Example:**
![Screenshot from 2024-06-13 16-09-06](https://github.com/AxisCommunications/axis-model-zoo/assets/28478841/35a054ae-6d3c-4d6b-8fc8-676713632c1a)
**Incorrect architecture Example:**
![Screenshot from 2024-06-13 16-12-45](https://github.com/AxisCommunications/axis-model-zoo/assets/28478841/f3f4f4ed-0731-49de-a696-8876b5860134)
If the architecture appears incorrect, the axis patch was not successfully applied, restart from step 1 of the guide and retrain your model.
* Make sure you are standing in the correct commit
* Make sure your model/yolov5n.yaml is patched and contains the line `activation: nn.ReLU6() `
* Make sure during the training you are adding the  `--cfg ` flag pointing to the right configuration file

3. Validate Model in PyTorch Format
Following step 4 of the tutorial on how to train the model, use the `val.py` script from the YOLOv5 repository with the `best.pt` weights. This step confirms the success of your training. If results are poor, check your data or extend training duration before proceeding.

4. Validate Model in TFLite Format
Again, following step 4, this time using the `val.py` script with `best-int8.tflite` weights. A performance drop of 0.01 to 0.03 mAP is expected. If the drop is larger, ensure `TensorFlow` version is between `2.4.0` and `2.13.1`. The axis patch enforces this during installation, so this should only be an issue if you skipped the pip install command. Note that `TensorFlow 2.13.2` has known issues, and newer versions are untested.

5. Local Model Testing
Before deploying to the camera, test your model locally with the detect.py script using both PyTorch and TFLite weights:
```bash
python3 detect.py --weights best.pt --source test_image.png
python3 detect.py --weights best-int8.tflite --source test_image.png
```
This repeats the validation of steps 3 and 4 on a specific image.

6. Test with larod-client
Before building the ACAP, use the larod-client to test your model. This CLI tool on the camera allows for quicker compatibility and speed tests. Refer to [this guide](https://developer.axis.com/computer-vision/computer-vision-on-device/test-your-model).

7. Compare larod-client and Local TensorFlow Outputs
Use `tf.lite.Interpreter` locally and `larod-client` on the device to test your model with a constant input image, then compare the binary outputs. Expect an average byte error of ~1-2 bits. If the error is larger, ensure correct model quantization per your chip (see [quantization documentation](https://developer.axis.com/computer-vision/computer-vision-on-device/quantization#quantization-for-each-dlpu)). For Axis OS >= 11.7, if quantization and previous tests are correct, but issues persist, contact support.

9. Make sure you are parsing correctly the model output.
If all previous steps are successful, but issues remain, the problem likely lies in parsing the model output in ACAP. Refer to the guide for further instructions on output parsing. [How to parse Yolov5 output](https://github.com/AxisCommunications/axis-model-zoo/discussions/45).
